### ISSUE TITLE
Feature request to support conditional directives

### ISSUE THREAD

**Issue by maettu-this** [oleg-shilo/cs-script#25]
Hi Oleg,

Another thing I encountered (related to issue #24):

Resources are built into \obj\Debug or \obj\Release or other directories, depending on the build configurations of a project. Thus, a script using these resources must be able to choose among the directories. This could be achieved as follows:

```
#if (DEBUG)
//css_resource ..\obj\Debug\MyProject.Properties.Resources.resources;
//css_resource ..\obj\Debug\MyProject.MyForm.resources;
#else
//css_resource ..\obj\Release\MyProject.Properties.Resources.resources;
//css_resource ..\obj\Release\MyProject.MyForm.resources;
#endif
```

Or:

```
//css_conditional_if(DEBUG)
    //css_resource ..\obj\Debug\MyProject.Properties.Resources.resources;
    //css_resource ..\obj\Debug\MyProject.MyForm.resources;
//css_conditional_else
    //css_resource ..\obj\Release\MyProject.Properties.Resources.resources;
    //css_resource ..\obj\Release\MyProject.MyForm.resources;
//css_conditional_endif
```

Or (following the .NET ConditionalAttribute):

```
//css_conditional(DEBUG) css_resource ..\obj\Debug\MyProject.Properties.Resources.resources;
//css_conditional(RELEASE) css_resource ..\obj\Release\MyProject.Properties.Resources.resources;

//css_conditional(DEBUG) css_resource ..\obj\Debug\MyProject.MyForm.resources;
//css_conditional(RELEASE) css_resource ..\obj\Release\MyProject.MyForm.resources;
```

Thanks & Best regards,
Matthias

**Comment by oleg-shilo** [comment]
It's not as simple as it seems. The `DEBUG` symbol is not available to the CS-Script engine. It is a compiler option that is passed to the actual C# compiler but not to CS-Script. 

The problem you are trying to solve is actually outside of the scripting domain. You are trying to make your script to behave as a Visual Studio project. While there is some definite resemblance certain VS concepts are just not applicable for scripting. For example project configuration `Debug` vs `Release`.

You can still have a conditional includes and references but you will need to have some external subsystem to do the condition evaluation:

```C#
//css_dir ..\obj\Debug 
//or css_dir ..\obj\Release
//css_resource MyProject.Properties.Resources.resources;
//css_resource MyProject.MyForm.resources;
```
Or you can set envvar form the launcher batch file `set MY_RESOURCES=..\obj\Debug` if batch file is appropriate for your case.
```C#
//css_resource %MY_RESOURCES%\MyProject.Properties.Resources.resources;
```
You can even set envvar from the pre-script (http://www.csscript.net/help/pre_post_scripts.html):
Pre-script:
```C#

using System;

class Script
{
    static public void Main(string[] args)
    {
        if (args.Contains("debug"))
            Environment.SetEnvironmentVariable("MY_RESOURCES", @"..\obj\Debug");
        else
            Environment.SetEnvironmentVariable("MY_RESOURCES", @"..\obj\Release");
    }
}
```
Primary script:
```C#
//css_pre setEV("debug");
//css_res %MY_RESOURCES%\MyProject.Properties.Resources.resources;
```
-------------
But all this becomes irrelevant when (very soon) your another feature request is implemented:
```C#
//css_res Properties/MyProject.Properties.Resources.resx;
//css_res MyProject.MyForm.resx;
```

**Comment by oleg-shilo** [comment]
Correct. The user indeed can pass "conditional compilation symbols" to the C# compiler (csc.exe) by using CS-Script directives and command line arguments. But they are just tunneled to the csc.exe and CS-Script never analyses them. However even if it did it wouldn't help as CS-Script cannot reliably know the all user supplied symbols at the stage of analysing the C# code for `//css_*` directives. 

The important point is that C# script execution is conducted in two phases. First is CS-Script quick/light code analysis and the second one is C# compiler compilation. The all compilation symbols are fully aggregated only after the first phase and it is already too late for the analysis algorithm to be altered.

Anyway, CS-Script analysis of hash-conditions in conjunction with collecting of compilation symbols is not trivial but not impossible. Though it has to be a business case for this feature to be implemented. Right now the original problem (#24 ) that triggered this proposal is solved rather directly so I don't see any serious trigger for the development to support this proposal. Sorry, it's a simple value vs. effort decision.

However I will keep this issue open for some time and will re-tag it as `enhancement`. If it is upvoted by the others then I see no problem of revisiting it again.

**Comment by maettu-this** [comment]
OK, I agree with the value vs. effort decision, it's fine with me.

**Comment by oleg-shilo** [comment]
Another interesting opportunity for conditional referencing is the use of [precompilers](http://www.csscript.net/help/precompilers.html). It is a technique that has replaced a pre-execution scripts. 

You can use a precompiler as below to conditionally add assembly references at runtime (e.g. System.Windows.Forms.dll):

`dynamic_ref.cs`
```C#
using System.Collections.Generic;
using System;
using System.Collections;

public class Precompiler
{
    public static bool Compile(ref string code, string scriptFile, bool isPrimaryScript, Hashtable context)
    {
        var newRefs = (List<string>)context["NewReferences"];
        newRefs.Add("System.Windows.Forms.dll");

        return false; //false as the code has not been modified
    }
}
```
`script.cs`
```C#
//css_pc dynamic_ref.cs;
using System;

class Script
{
    static public void Main(string[] args)
    {
        System.Windows.Forms.MessageBox.Show("ttt");
    }
}
```

And inside of a precompiler you can analyse the compiler options of the primary script with `(string)context["CompilerOptions"]`.

**Comment by maettu-this** [comment]
I noticed that you cleaned up and closed several of the "old" issues, and I understand that issues that require significant efforts more likely get rejected if nobody else wants it. Still, for this issue, I would like to give it another try. Instead of the preprocessor based approach...
```
#if (DEBUG)
//css_reference .\Debug\MyAssembly.dll;
#else
//css_reference .\Release\MyAssembly.dll;
#endif
```
...which definitely needs quite some additional parsing, hence lots of efforts. How about an attribute-like approach...
```
//css_reference -conditional(DEBUG) .\Debug\MyAssembly.dll;
//css_reference -conditional(RELEASE) .\Release\MyAssembly.dll;
```
? While not as catchy as the first approach, the second approach would give the same result, at much less cost for parsing. Any chance for this?

**Comment by maettu-this** [comment]
Well, in case of my hosting application I pass compiler definitions anyway, since I use `CSScript.CompileWithConfig()`.

Note that I proposed `conditional` rather than `if` for similarity with the .NET `ConditionalAttribute`. But of course `if` is equally obvious.

**Comment by oleg-shilo** [comment]
While thinking more about it I started leaning towards using envars as a unified data channel. This way all conditional symbols from all sources (CLI, //css_co, CompileWithConfig) would be tunneled to the parses as envars. This would allow simple and decoupled implementation. The only side effect of this approach is that both symbols and envars will be defined with the same syntax. But it is a small price to pay:

```C#
//css_dir #if DEV_PC .\temp\cots
//css_dir #if DEBUG .\Debug
```

**Comment by maettu-this** [comment]
That's indeed a small price to pay 👍

**Comment by maettu-this** [comment]
Hi Oleg, I am back ;-)

I am upgrading my hosting application and of course have upgraded to the latest revision of CS-Script. Conditional compilation now is available, thus trying to make use of it where I need it.

While it works in principle, I have found some issues. I am listing them here. If you prefer, I can move them to a new issue.

### Sequence (major)
It seems as if the preprocessing only works **after** the script could successfully be compiled. But this won't work if the script imports or references necessary code.

Situation:
```
 > Script.cs
    > Debug\Library.dll
    > Release\Library.dll
```
Now I expected this to work in `Script.cs`:
```
//css_reference #if (DEBUG) .\Debug\Library.dll;
//css_reference #if (!DEBUG) .\Release\Library.dll;
...
Library.DoSomething();
``` 
`error CS0103: The name 'Library' does not exist in the current context`

Unconditionally works fine:
```
//css_reference .\Debug\Library.dll;
...
Library.DoSomething();
``` 

Using `css_dir` doesn't work either
```
//css_dir #if (DEBUG) .\Debug
//css_dir #if (!DEBUG) .\Release
//css_reference Library.dll;
...
Library.DoSomething();
``` 
`(0,0): error CS0006: Metadata file 'Library.dll' could not be found`

### Error location (minor)
The above error message erroneously states (0,0) instead of the line of the code where 'Library.dll' is referenced.

### Error message (minor)
`"Cannot parse \"//css_reference\nEnsure your directive escapes all delimiters (';(){},') by doubling the delimiter character."`
Even though I am running on Windows, the error message uses a non-Windows `\n`. Thus splitting the string for `Environment.NewLine` will not work. I propose to always use `Environment.NewLine` when composing multi-line strings.

**Comment by oleg-shilo** [comment]
It's close to impossible to understand the details of your problem as you are describing the outcome but not what you are doing.

Anyway, I have composed the sample for you that shows the technique. See attached file.

After experimenting with your scenario I am convinced that on practical level it is more convenient to use envars instead of condition in the `//css_` directive.

```C#
//css_ref .\%BUILD_TYPE%\Library.dll
``` 
Where `BUILD_TYPE` is the environment variable with value either "Release" or "Debug".

[ConsoleApp1.zip](https://github.com/oleg-shilo/cs-script/files/5429950/ConsoleApp1.zip)

**Comment by maettu-this** [comment]
Same here, for a system where everything is known, envars are fine. But then I will have to tell my users that `%BUILD_TYPE%` exists and what it means. But users are used to define symbols in an e.g. Visual Studio project definition. But now I think I got it, I have to pass all defined symbols as env vars. Somehow didn't get that from what I read further above.

This will solve the first of the four issues. Is it OK to handle the other issues here or do you want me to file three separate issue items for them?

**Comment by maettu-this** [comment]
Asking again, is it OK to handle the other issues here or do you want me to file three separate issue items for them? The three others are tightly related to this one, so for me it's pragmatic to handle them here. However, developers have different opinions on grouping issues. So I ask about your preference in this case.

**Comment by oleg-shilo** [comment]
It is better to have an issue item per problem.

> But then I will have to tell my users that %BUILD_TYPE% exists and what it means.

You will always have to communicate to your users the existence of the conditional symbol. Your app is playing the same role as the IDE (VS) so it needs to collect user defined symbols and tunnel them via envars.

It is possible to change the script engine to do the tunneling automatically and it is a simple change. But I am not sure that polluting envars even for a good but not obvious reason (like yours) makes sense. Some severe side effects are possible.

Envars stack has limit so I only want to overload it if there is really good use-case for that.

**Comment by maettu-this** [comment]
Related issues moved to new #212, #213, #214.

Thanks for providing this feature and the explanations given. It's not always obvious on what exactly has to be done, but this is in the nature of engineering ;-)
